pynav/batch.py

```diff
@@ -0,0 +1,313 @@
+        e = self._y.value.reshape((-1, 1)) - y_check.reshape((-1, 1))
+
+        # Weight error by square root of information matrix
+        L = self._y.model.sqrt_information(x)
+        e = L.T @ e
+
+        if compute_jacobians:
+            jacobians = [None] * len(states)
+
+            if compute_jacobians[0]:
+                jacobians[0] = -L.T @ self._y.model.jacobian(x)
+            return e, jacobians
+
+        return e
+
+
+@dataclass
+class OptimizationSettings:
+    """A dataclass to hold some default optimizaiton settings."""
+    solver: str = "GN" # Solver type, either "GN" or "LM"
+    max_iters: int = 100 # maximum number of iters
+    step_tol: float = 1e-7 # Convergence tolerance, norm of the step size.
+    tau: float = 1e-11 # tau parameter in LM
+    verbose: bool = True # Print convergence during runtime 
+
+
+def run_batch(
+    x0: State,
+    P0: np.ndarray,
+    input_data: List[Input],
+    meas_data: List[Measurement],
+    process_model: ProcessModel,
+    return_opt_results: bool = False,
+    opt_settings: OptimizationSettings = OptimizationSettings(),
+) -> List[StateWithCovariance]:
+    """Creates and solves a batch problem using pysquares.
+
+    The input data is used to propagate the initial state x0 forward in time
+    using the process model, to generate an initial estimate of the state
+    at estimate timestep.
+
+    The batch problem created involves a PriorResidual, a ProcessResidual 
+    for each input used to connect subsequent states through the process model,
+    and MeasurementResiduals for each measurement. 
+
+    Parameters
+    ----------
+    x0 : State
+        x0: Initial state.
+    P0 : np.ndarray
+        Initial covariance
+    input_data : List[Input]
+        List of input data.
+    meas_data : List[Measurement]
+        List of measurements.
+    process_model : ProcessModel
+        Process model used to propagate the initial estimate
+        and form ProcessResiduals.
+    return_opt_results : bool, optional
+        Flag to optionally return the results dictionary
+        from the batch problem, by default False
+    opt_settings : OptimizationSettings, optional
+        Settings used by the optimizer, by default OptimizationSettings()
+
+    Returns
+    -------
+    List[StateWithCovariance]
+        List of estimates with covariance.
+    """
+
+    # Sort the data by time
+    input_data.sort(key=lambda x: x.stamp)
+    meas_data.sort(key=lambda x: x.stamp)
+
+    # Remove all that are before the current time
+    for idx, u in enumerate(input_data):
+        if u.stamp >= x0.stamp:
+            input_data = input_data[idx:]
+            break
+
+    for idx, y in enumerate(meas_data):
+        if y.stamp >= x0.stamp:
+            meas_data = meas_data[idx:]
+            break
+
+    # We want to generate state estimates at
+    # each input and measurement timestamp
+    input_stamps = [round(u.stamp, 5) for u in input_data]
```

Comment:
we should discuss this. i guess a `dt=1e-5` is plenty small, but why is this necessary for generality's sake?